### ISSUE TITLE
nn.Linear weight initalization - uniform or kaiming_uniform?

### ISSUE THREAD

**Issue by adrianstaniec** [pytorch/pytorch#57109]
IMHO there is a discrepancy between the docs and code of nn.Linear, when it comes to initialization.

documentation says that the weights are initialized from
uniform ( 1/sqrt(in_ feaures) , 1/sqrt(in_ feaures)):
https://github.com/pytorch/pytorch/blob/0df574017d8f0eb1527e973defb9c65a5befe966/torch/nn/modules/linear.py#L53-L56

code says that the weights are initialized from
kaiming_uniform 
https://github.com/pytorch/pytorch/blob/77721ee318d6785010144aa4569efb98199e7162/torch/nn/modules/linear.py#L88-L89
and that includes factors of sqrt(3), gain based on 'a', and 'fan':
https://github.com/pytorch/pytorch/blob/77721ee318d6785010144aa4569efb98199e7162/torch/nn/init.py#L390-L395

Is that an error or am I missing something?

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @brianjo @mruberry @albanD

**Comment by zou3519** [comment]
Hi-pri because the docs say something different from the code. We should probably document the factor of sqrt(3) that we use.

**Comment by soumith** [comment]
the `kaiming_init` is used for convenience, but basically the sqrt(5) goes in and gets simplified in the formula for `gain` as `sqrt(2 / (sqrt(5) * sqrt(5))` which is `sqrt(1/3)`.

The overall logic simplifies to this equivalent logic:

https://github.com/pytorch/pytorch/blob/v0.4.1/torch/nn/modules/conv.py#L40-L47

**Comment by soumith** [comment]
relevant PR: https://github.com/pytorch/pytorch/pull/9038#pullrequestreview-134370519

**Comment by soumith** [comment]
> We should probably document the factor of sqrt(3) that we use. I remember that there was some justification for it

Uniform distribution's stdv is basically multiplied by sqrt(3) when you draw: https://physics.nist.gov/cuu/Uncertainty/typeb.html and https://physics.stackexchange.com/questions/110242/why-is-uncertainty-divided-by-sqrt3

If you notice in my "equivalent logic" [snippet](https://github.com/pytorch/pytorch/blob/v0.4.1/torch/nn/modules/linear.py#L48-L52), one thing worth noticing here is that by default, we don't multiply the stdv with sqrt(3).
The reason we don't do that because Collobert at al. in some historical past have figured out that this not-multiplying and having a slight gain in the uniform distributions heuristically works better. This is not recorded in literature anywhere but has been recorded in code since Lush, Senna, Torch5, Torch7 and now PyTorch which is somewhat unfortunate. A detailed discussion of this was recorded on Google Plus, which is now defunct and that discussion has been erased from the internet. However, I've revived a copy of that discussion here: https://soumith.ch/files/20141213_gplus_nninit_discussion

**Comment by ciaochiaociao** [comment]
> the `kaiming_init` is used for convenience, but basically the sqrt(5) goes in and gets simplified in the formula for `gain` as `sqrt(2 / (sqrt(5) * sqrt(5))` which is `sqrt(1/3)`.
> 
> The overall logic simplifies to this equivalent logic:
> 
> https://github.com/pytorch/pytorch/blob/v0.4.1/torch/nn/modules/linear.py#L48-L52

Just for others coming to this issue, there's a typo, which is "1" missing in sqrt(2 / 1 + (sqrt(5) * sqrt(5)) as in https://pytorch.org/docs/stable/nn.init.html

**Comment by Ehliu** [comment]
They're the same

**Comment by redwrasse** [comment]
Reviving this discussion since it's related to understanding a good default weights initialization, or at least a consistent one, for the sister module, `torch.nn.Bilinear`: https://github.com/pytorch/pytorch/issues/132231

I agree with @matthijsvk, the code is confusing and better just to set the uniform distribution with bounds directly.
And, a note about the missing sqrt(3) factor would be nice- the Glorot initialization paper https://proceedings.mlr.press/v9/glorot10a/glorot10a.pdf in equation (13) argues trying to set ` (n * Var(W))^(d-i)` to `1` to control variance of gradients wrt earlier layers, hence a multiplicative root 3 factor to the bound.

**Comment by max0x7ba** [comment]
`Linear.reset_parameters` does `init.kaiming_uniform_(self.weight, a=math.sqrt(5))`, whereas parameter `a` is the negative slope for `LeakyReLU` and its default is 0.01. 

Why does it use constant `math.sqrt(5)` regardless of actual negative slope for `LeakyReLU`, please?

**Comment by pollo-coder** [comment]
> The source code should be clear and understanable.
> 
> Right now the code really does this:
> 
> ```
> fan_in, _ = nn.init._calculate_fan_in_and_fan_out(m.weight)
> # generate sqrt(3) in complicated way: sqrt(2 / (1 + 5)) = sqrt(1/3)
> gain = torch.nn.init.calculate_gain("leaky_relu", math.sqrt(5)) 
> # sqrt(3) b/c uniform gets rid of the sqrt(1/3) again
> bound = math.sqrt(3) * gain / math.sqrt(fan_in) 
> nn.init.uniform_(weight, -bound, bound)
> ```
> 
> Much more clearer would be:
> 
> ```
> fan_in, _ = nn.init._calculate_fan_in_and_fan_out(m.weight)
> bound = 1 / math.sqrt(fan_in)
> nn.init.uniform_(weight, -bound, bound)
> ```
> 
> And an explanation why we don't actually use Kaiming He's inititialization but values that are sqrt(3)/sqrt(6) smaller (with gain from 'linear'/'relu'), even though the user may think that from reading the source code.

Of course - what is the reason why this cannot be changed as you suggest?
